lib/resty/dns/resolver.lua

```diff
@@ -398,11 +414,38 @@ local function parse_response(buf, id)
         answers.errstr = resolver_errstrs[code] or "unknown"
     end
 
-    for i = 1, nan do
+    local nan_nns = nan + nns
+    local nan_nar = nan_nns + nar
+
+    local authority_section = false
+    local additional_section = false
+
+    if opts then
+        authority_section = opts.authority_section
+        additional_section = opts.additional_section
+    end
+
+    for i = 1, nan_nar do
         -- print(format("ans %d: qtype:%d qclass:%d", i, qtype, qclass))
+        local section
+        local should_skip = false
+
+        if i <= nan then
+            section = SECTION_AN
+        elseif i > nan and i <= nan_nns then
+            section = SECTION_NS
+            should_skip = not authority_section
+        elseif i > nan_nns and i <= nan_nar then
+            section = SECTION_AR
+            should_skip = not additional_section
+        end
 
         local ans = {}
-        insert(answers, ans)
+        ans.section = section
+
+        if not should_skip then
```

Comment:
Instead of always parsing the unwanted sections, we should just stop parsing directly?